tests/unit/encode_test.py

```diff
@@ -76,3 +81,183 @@ def test_encode_abi():
+    aats_out_is_none = dre.abi_argument_types is None
+
+    if aats_out_is_none:
+        assert dre.method is None
+        assert dre.selector is None
+        assert dre.abi_method_signature is None  # so will skip _abi_adapter() call
+        assert dre.abi_argument_types is None  # so will encode args without
+
+        if not args_below_max_num:
+            with pytest.raises(AssertionError) as ae:
+                dre._executor_prep(args)
+
+            assert (
+                "for non-ABI app calls, there is no specification for encoding more than"
+                in str(ae.value)
+            )
+            return
+
+        encoded_args = DryRunEncoder.encode_args(args)
+        args_out, encoded_args_out = dre._executor_prep(args)
+        assert args_out == args
+        assert encoded_args_out == encoded_args
+        return
+
+    assert dre.method
+    assert dre.selector
+    assert dre.abi_method_signature
+    assert isinstance(dre.abi_argument_types, list)
+
+    argnum_same = len(args) == len(dre.abi_argument_types)
+    argnum_one_more = len(args) == len(dre.abi_argument_types) + 1
+    arg0_is_selector = args and args[0] == dre.selector
+
+    if validation:
+        if argnum_one_more:
+            if not arg0_is_selector:
+                with pytest.raises(AssertionError) as ae:
+                    dre._executor_prep(args)
+
+                assert "should have been the selector" in str(ae.value)
+                return
+
+        if not (argnum_same or argnum_one_more):
+            with pytest.raises(AssertionError) as ae:
+                dre._executor_prep(args)
+
+            assert "is incompatible with" in str(ae.value)
+            return
+
+        try:
+            prefix = tuple() if omit_method_selector else ("blah",)
+            type_prefix = [] if omit_method_selector else [None]
+            encoded_args = DryRunEncoder.encode_args(
+                prefix + args,
+                type_prefix + dre.abi_argument_types,
+                validation=validation,
+            )
+        except AssertionError as encode_args_ae:
+            with pytest.raises(AssertionError) as ae:
+                dre._executor_prep(args)
+            assert "problem encoding arg" in str(encode_args_ae)
+            assert str(ae.value) == str(encode_args_ae)
+            return
+
+        args_out, encoded_args_out = dre._executor_prep(args)
+        argslen = len(args)
+        aolen = len(args_out)
+        arg_range = slice(aolen - argslen, aolen)
+        assert args_out[arg_range] == args
+        assert len(encoded_args_out) == len(encoded_args)
+        assert encoded_args_out[arg_range] == encoded_args[arg_range]
+        return
+
+    assert validation is False
+    assert aats_out_is_none is False, "already considered this"
+
+    try:
+        encoded_args = DryRunEncoder.encode_args(
+            args[1 - int(omit_method_selector) :],
+            dre.abi_argument_types[1 - int(omit_method_selector) :],
```

Comment:
nit: I think it would be better to introduce a variable `start_from: int = 1 - int(omit_method_selector)` here